File: 3sem/4lab/Banks/Entities/Bank.cs
using Banks.BankAccounts;
using Banks.BankAccounts.Factories;
using Banks.Clocks;
using Banks.Exceptions;
using Banks.Models.BankConfigurations;
using Banks.Services;

namespace Banks.Entities;

public class Bank
{
    private readonly HashSet<Client> _clients;
    private readonly HashSet<IBankAccount> _accounts;

    public Bank(string name, BankConfiguration configuration, CentralBank centralBank)
    {
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(configuration);
        ArgumentNullException.ThrowIfNull(centralBank);

        Name = name;
        Configuration = configuration;
        CentralBank = centralBank;

        Id = Guid.NewGuid();
        _clients = new HashSet<Client>();
        _accounts = new HashSet<IBankAccount>();
    }

    public Guid Id { get; }
    public string Name { get; }
    public BankConfiguration Configuration { get; }
    public CentralBank CentralBank { get; }

    public IEnumerable<Client> Clients => _clients;
    public IEnumerable<IBankAccount> Accounts => _accounts;

    public void AddClient(Client client)
    {
        ArgumentNullException.ThrowIfNull(client);

        if (!_clients.Add(client))
            throw InvalidBankOperationException.OnAddExistingClient();
    }

    public IBankAccount CreateAccount(IBankAccountFactory factory, Client client, IClock clock)
    {
        ArgumentNullException.ThrowIfNull(factory);
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(clock);

        if (!_clients.Contains(client))
            throw InvalidBankOperationException.OnAddAccountToUnregisteredClient();

        IBankAccount account = factory.Create(this, client, clock);
        _accounts.Add(account);

        return account;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/Entities/Bank.java
package BanksRemake.Domain.Entities;

import BanksRemake.Domain.BankAccounts.BankAccount;
import BanksRemake.Domain.BankAccounts.Factories.BankAccountFactory;
import BanksRemake.Domain.Clocks.Clock;
import BanksRemake.Domain.Exceptions.InvalidBankAccountOperationException;
import BanksRemake.Domain.Exceptions.InvalidBankOperationException;
import BanksRemake.Domain.Models.BankConfigurations.BankConfiguration;
import BanksRemake.Domain.Services.CentralBank;
import lombok.Getter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * Bank serving clients.
 */
public class Bank {
    private final ArrayList<Client> clients;
    private final ArrayList<BankAccount> accounts;

    @Getter
    private final UUID id;
    @Getter
    private final String name;
    @Getter
    private final BankConfiguration configuration;
    @Getter
    private final CentralBank centralBank;

    public Bank(String name, BankConfiguration configuration, CentralBank centralBank) {
        this.name = name;
        this.configuration = configuration;
        this.centralBank = centralBank;

        this.id = UUID.randomUUID();
        this.clients = new ArrayList<>();
        this.accounts = new ArrayList<>();
    }

    public List<Client> getClients() {
        return Collections.unmodifiableList(clients);
    }

    public List<BankAccount> getAccounts() {
        return Collections.unmodifiableList(accounts);
    }

    /**
     * Adding a client to the bank.
     * @param client unregistered client
     */
    public void addClient(Client client) throws InvalidBankOperationException {
        if (clients.contains(client)) {
            throw InvalidBankOperationException.onAddExistingClient();
        }

        clients.add(client);
    }

    /**
     * Creating a bank account for a client.
     * @param factory abstract factory that creates different types of bank accounts.
     * @param client registered client.
     * @param clock a clock that notifies accounts of the arrival of a new day or month.
     * @return created bank account.
     */
    public BankAccount createAccount(BankAccountFactory factory, Client client, Clock clock) throws InvalidBankOperationException, InvalidBankAccountOperationException {
        if (!clients.contains(client)) {
            throw InvalidBankOperationException.onAddAccountToUnregisteredClient();
        }

        BankAccount account = factory.create(this, client, clock);
        accounts.add(account);

        return account;
    }
}
